src/hip_memory.cpp

```diff
@@ -1161,42 +1161,56 @@ hipError_t hipMemcpy3D(const struct hipMemcpy3DParms *p)
     return ihipLogStatus(e);
 }
 
-// TODO - make member function of stream?
-template <typename T>
-void
-ihipMemsetKernel(hipStream_t stream,
-    LockedAccessor_StreamCrit_t &crit,
-    T * ptr, T val, size_t sizeBytes,
-    hc::completion_future *cf)
+namespace
 {
-    int wg = std::min((unsigned)8, stream->getDevice()->_computeUnits);
-    const int threads_per_wg = 256;
+    template<
+        uint32_t block_dim,
+        typename RandomAccessIterator,
+        typename N,
+        typename T>
+    __global__
+    void hip_fill_n(RandomAccessIterator f, N n, T value)
+    {
+        const uint32_t grid_dim = hipGridDim_x;
 
-    int threads = wg * threads_per_wg;
-    if (threads > sizeBytes) {
-        threads = ((sizeBytes + threads_per_wg - 1) / threads_per_wg) * threads_per_wg;
+        size_t idx = hipBlockIdx_x * block_dim + hipThreadIdx_x;
+        while (idx < n) {
+            new (&f[idx]) T{value};
```

Comment:
This is just a simple assignment correct?  What is motivation to use placement new here - especially inside device kernel?